### ISSUE TITLE
SVC: missing trust certificate when authenticating OpenShift social login

### ISSUE THREAD

**Issue by meiaus** [OpenLiberty/open-liberty#9881]
Problem scenario:
The user application (ACME) deployed in OCP 4.2 with social login feature (OAuth with Open Shift) didn't supply the trust certificate (as used in server.env `cert_defaultKeyStore=/var/run/secrets/kubernetes.io/serviceaccount/ca.crt`), when attempting to get the authentication after hitting the application URL (e.g. https://acme-proxy-acmetest.apps.kneeing.os.fyre.ibm.com/AcmeAnnuityWeb), the failure appeared as in browser:
```
HTTP Error 401
Error: CWWKS5489E: An error was encountered while authenticating a user by using social media.
``` 
as well as messages.log:

```
[11/20/19, 15:39:46:477 GMT] 00000037 com.ibm.ws.logging.internal.impl.IncidentImpl                I FFDC1015I: An FFDC Incident has been created: "sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target com.ibm.ws.ssl.core.WSX509TrustManager checkServerTrusted" at ffdc_19.11.20_15.39.46.0.log
[11/20/19, 15:39:46:478 GMT] 00000037 com.ibm.ws.ssl.core.WSX509TrustManager                       E CWPKI0823E: SSL HANDSHAKE FAILURE:  A signer with SubjectDN [CN=172.30.0.1] was sent from the host [kubernetes.default.svc:443].  The signer might need to be added to local trust store [/opt/ol/wlp/output/acme_server1/resources/security/key.p12], located in SSL configuration alias [defaultSSLConfig].  The extended error message from the SSL handshake exception is: [PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target].
[11/20/19, 15:39:46:499 GMT] 00000037 com.ibm.ws.logging.internal.impl.IncidentImpl                I FFDC1015I: An FFDC Incident has been created: "javax.net.ssl.SSLHandshakeException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target com.ibm.ws.security.social.internal.utils.OpenShiftUserApiUtils 60" at ffdc_19.11.20_15.39.46.1.log
[11/20/19, 15:39:46:686 GMT] 00000037 com.ibm.ws.security.social.tai.TAIUserApiUtils               W CWWKS5461E: The social login feature encountered an error while getting user information from the user API [https://kubernetes.default.svc/apis/authentication.k8s.io/v1/tokenreviews] that is configured for social login configuration [openshiftLogin]. CWWKS5371E: The response from the OpenShift user API cannot be processed: javax.net.ssl.SSLHandshakeException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
[11/20/19, 15:39:46:708 GMT] 00000037 com.ibm.ws.security.social.error.SocialLoginException        E CWWKS5452E: The social login feature cannot authenticate the user because the response from the user API that is configured for the social login configuration [openshiftLogin] is null or empty.
[11/20/19, 15:39:46:939 GMT] 00000037 com.ibm.ws.logging.internal.impl.IncidentImpl                I FFDC1015I: An FFDC Incident has been created: "com.ibm.ws.security.social.error.SocialLoginException: CWWKS5452E: The social login feature cannot authenticate the user because the response from the user API that is configured for the social login configuration [openshiftLogin] is null or empty. com.ibm.ws.security.social.tai.OAuthLoginFlow 98" at ffdc_19.11.20_15.39.46.2.log

```
[social_login_logs.zip](https://github.com/OpenLiberty/open-liberty/files/3870177/social_login_logs.zip)
- the zip contains:
```
-rw-r--r-- 1 root root  134649 Nov 20 11:04 ffdc_19.11.20_15.39.46.2.log
-rw-r--r-- 1 root root   17174 Nov 20 11:04 ffdc_19.11.20_15.39.46.1.log
-rw-r--r-- 1 root root   16330 Nov 20 11:04 ffdc_19.11.20_15.39.46.0.log
-rw-r--r-- 1 root root 1110311 Nov 20 11:01 server_trace.log
-rw-r--r-- 1 root root   25704 Nov 20 11:01 messages.log
```

**Diagnostic information:**

```
********************************************************************************
product = Open Liberty 19.0.0.12 (wlp-1.0.35.cl191220191118-0810)
wlp.install.dir = /opt/ol/wlp/
server.output.dir = /opt/ol/wlp/output/acme_server1/
java.home = /opt/java/openjdk
java.version = 11.0.5
java.runtime = OpenJDK Runtime Environment (11.0.5+10)
os = Linux (4.18.0-80.11.2.el8_0.x86_64; amd64) (en_US)

```

Questions or considerations:
1. To help user provide the valid certificate in this scenario, will the message be added to show the information, or can only be done via the documentation as a hint to the user?
2. The messages also contain the user API social login exceptions, which may be too much information in this scenario? Although this related error may go away once the trust issue is resolved, wondering if it can be suppressed in this case to avoid additional confusion?

**Comment by ayoho** [comment]
1. Adding the appropriate certificates to the right spot is something that's a much larger scope than what we'd be able to, or want to, include in a single NLS message. So the message that's emitted by the SSL code is about all we're willing to do - the rest will be up to the server admin to check documentation to see what they might need to do to address the problem.
1. The user API exceptions in the social login code are actually just a bunch of catch blocks behind handled as the SSL handshake error gets propagated back up through the stack. The request and its processing is really done. What you're seeing is each level of the call stack getting its chance to log the failure and emit any FFDC for whatever exception it catches.

**Comment by meiaus** [comment]
@ayoho thanks for your feedback.  I guess there is not much to be done from messages point of view. But I'd like to confirm with @brutif for doc (or blog) as preventive approach.

@brutif, in the blog or other form of doc, will this be added as a hint to handle signer certificate before deploying to OpenShift? although this is not social login feature's scope, if this is not handled correctly, the application can already fail before using the social login feature.   Specifically, if user tries to deploy their existing 'non-OpenShift' application and with the hint, they have a  chance to review their configuration and be prepared for failure if they haven't done the modification.

**Comment by brutif** [comment]
@meiaus they do get a message telling them what they need to do -
```
CWPKI0823E SSL HANDSHAKE FAILURE:  A signer with SubjectDN [CN=172.30.0.1] was sent from the host [kubernetes.default.svc:443].  The signer might need to be added to local trust store 
```
We probably won't be able document the entire process for setting up with openshift in the server.xml example.  We could see if anything can be done to improve the CWPKI0823E message.

**Comment by meiaus** [comment]
Thanks @brutif for the feedback.

Checked with security development about the new feature (to be delivered in 19.0.0.12) "Reading a certificate from an environment variable", which is described in the draft doc for the usage of `cert_defaultKeyStore` as leveraged in this Social Login scenario.

Closing this ticket and will check back when this doc is published for 19.0.0.12.
